token-metadata/js/src/instruction.ts

```diff
@@ -0,0 +1,56 @@
+import type { PublicKey } from '@solana/web3.js';
+
+import { Field, TokenMetadata } from './state';
+
+
+/*
+Not implimented as a strict interface, by function signatures so that the 
+library can work with programs that implement the Rust interface.
+*/
+
+
+// TODO:- Library should impliment Pack/unpack helpers
+export type Pack = (meta: TokenMetadata) => Promise<Buffer>
+export type Unpack = (input: Buffer) => Promise<TokenMetadata>
```

Comment:
So, you've actually got two pack-unpack operations in this interface. One is for the instructions, and the other is for the state.
So, as a user of this library I'll need to do both:
- Create instructions to send to an SPL Token Metadata program
- Deserialize account data to read SPL Token Metadata from a metadata account

Some of this work has been done already in the JavaScript TLV library, so you should just need to import it and drive it's unpacking functions to do the _state_ part of things.

https://github.com/solana-labs/solana-program-library/tree/master/libraries/type-length-value/js

For instructions, we're just building a `TransactionInstruction` and making sure we build the bytes the same way as we do in the Rust lib, so we get the right instructions!

Reply:
Thanks for the explanation, its getting a bit clearer to me, but still a bit unsure where to start.

Do you know a good way to get example byte data to build a test case around?

Reply:
For the instructions, you'll want to take note of the 8-byte discriminator for each:

https://github.com/solana-labs/solana-program-library/blob/f76276e91df169d045d5c9c1b240c1923bdd72e3/token-metadata/interface/src/instruction.rs#L22

Then it's just unpacking the rest with Borsh:

https://github.com/solana-labs/solana-program-library/blob/f76276e91df169d045d5c9c1b240c1923bdd72e3/token-metadata/interface/src/instruction.rs#L172-L175

For the state, there's one discriminator for token metadata state:

https://github.com/solana-labs/solana-program-library/blob/f76276e91df169d045d5c9c1b240c1923bdd72e3/token-metadata/interface/src/state.rs#L45

And the rest is also unpacked with Borsh:

https://github.com/solana-labs/solana-program-library/blob/f76276e91df169d045d5c9c1b240c1923bdd72e3/token-metadata/interface/src/state.rs#L104

The thing to keep in mind about the Token Metadata state is the fact that it can exist alongside other account data, such as a mint with extensions!

For this reason, it's unpacked with Borsh, but within a Type-Length-Value configuration, meaning the Token Metadata is one entry of some arbitrary possible number of entries, each of which are discriminated using an 8-byte `ArrayDiscriminator`.

As I mentioned, you should be able to use the linked JS library to unpack TLV, and as far as creating bytes for test cases, you might have to roll some buffers manually unfortunately!

Reply:
Thanks for the directions - I think i managed to figure out with your help. 

I've implemented the Instructions such that they return TransactionInstructions with the right data bytes layout.

I managed to cross check against the rust by running the code locally and printing some of the buffers